libs/cardano-data/src/Data/OSet/Strict.hs

```diff
@@ -0,0 +1,209 @@
+  (|>),
+  (<|),
+  (|><),
+  (><|),
+)
+where
+
+import Cardano.Ledger.Binary (DecCBOR (decCBOR), EncCBOR (encCBOR))
+import Cardano.Ledger.TreeDiff (ToExpr)
+import Control.DeepSeq (NFData)
+import Data.Foldable (foldl')
+import Data.Sequence.Strict qualified as SSeq
+import Data.Set qualified as Set
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks)
+import Prelude hiding (lookup, null, seq)
+
+-- | A general-purpose finite, ordered, set that is strict in its
+-- values.
+--
+-- The strictness is enforced by the underlying `StrictSeq` from base,
+-- and the uniqueness of the values is enforced in this module, by the
+-- constructing functions by leveraging an accompanying `Set`.
+data OSet a = OSet
+  { osSSeq :: !(SSeq.StrictSeq a)
+  , osSet :: !(Set.Set a)
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (ToExpr, NoThunks, NFData)
+
+instance EncCBOR a => EncCBOR (OSet a) where
+  encCBOR = encCBOR . toStrictSeq
+
+instance (Show a, Ord a, DecCBOR a) => DecCBOR (OSet a) where
+  decCBOR =
+    let failOnDuplicates (duplicates, oset) =
+          if Set.null duplicates
+            then pure oset
+            else fail $ "Duplicates found while decoding OSet: " <> show duplicates
+     in failOnDuplicates . fromStrictSeqDuplicates =<< decCBOR
+
+-- | \( O(1) \). Shallow invariant using just `length` and `size`.
+invariantHolds :: OSet a -> Bool
+invariantHolds (OSet seq set) = SSeq.length seq == Set.size set
+
+-- | \( O(n \log n) \). Deep invariant using set membership check for
+-- each value. By the pigeon-hole principle, this check is exhaustive.
+invariantHolds' :: Ord a => OSet a -> Bool
+invariantHolds' oset@(OSet seq set) = invariantHolds oset && all (`Set.member` set) seq
+
+-- | \( O(1) \).
+null :: OSet a -> Bool
+null (OSet _ set) = Set.null set
+
+-- | \( O(1) \).
+size :: OSet a -> Int
+size (OSet _ set) = Set.size set
+
+-- | \( O(1) \).
+empty :: OSet a
+empty = OSet SSeq.Empty Set.empty
+
+-- | \( O(1) \). Strict in its argument.
+singleton :: a -> OSet a
+singleton !x = OSet (SSeq.singleton x) (Set.singleton x)
+
+-- | \( O(\log(\min(i,n-i))) \). The element at the specified position,
+-- counting from 0. If the specified position is negative or at least
+-- the length of the sequence, 'lookup' returns 'Nothing'.
+lookup :: Int -> OSet a -> Maybe a
+lookup i (OSet seq _set) = SSeq.lookup i seq
+
+-- | `flip`ed version of `lookup`
+(!?) :: OSet a -> Int -> Maybe a
+(!?) = flip lookup
+
+-- | \( O(\log n) \). Checks membership before cons'ing.
+cons :: Ord a => a -> OSet a -> OSet a
+cons x oset@(OSet seq set) =
+  if x `Set.member` set
+    then oset
+    else OSet (x SSeq.<| seq) (x `Set.insert` set)
+
+(<|) :: Ord a => a -> OSet a -> OSet a
+(<|) = cons
+
+infixr 5 <|
+
+-- | \( O(\log n) \). Checks membership before snoc'ing.
+snoc :: Ord a => OSet a -> a -> OSet a
+snoc oset@(OSet seq set) x =
+  if x `Set.member` set
+    then oset
+    else OSet (seq SSeq.|> x) (x `Set.insert` set)
```

Comment:
you can keep it like it is, but I'd do it this way
```suggestion
snoc oset@(OSet seq set) x 
  | x `Set.member` set = oset
  | otherwise =  OSet (seq SSeq.|> x) (x `Set.insert` set)
```

Reply:
I like your suggestion! Thanks! Will change it. :+1: